### PR TITLE
Fix cyclic dependency and improve import time

### PR DESCRIPTION
**Pull request by kaxil**
This PR/commit improves the import time and removes cyclic dependency.

Reduces the import time for `astro/databases/__init__.py` from **23.254** seconds to **0.062** seconds

**Before**:

```
Thu Oct  6 23:12:29 2022    /Users/kaxilnaik/Documents/GitHub/astronomer/astro-sdk/python-sdk/dev/dags/program.prof

         3726676 function calls (3646580 primitive calls) in 23.254 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   4194/1    0.399    0.000   23.281   23.281 {built-in method builtins.exec}
3899/1    0.109    0.000   23.280   23.280 <frozen importlib._bootstrap>:1002([_find_and_load](http://127.0.0.1:4000/?func_name=_find_and_load))
   3849/1    0.074    0.000   23.280   23.280 <frozen importlib._bootstrap>:967([_find_and_load_unlocked](http://127.0.0.1:4000/?func_name=_find_and_load_unlocked))
   4500/2    0.023    0.000   23.276   11.638 <frozen importlib._bootstrap>:220([_call_with_frames_removed](http://127.0.0.1:4000/?func_name=_call_with_frames_removed))
   3672/2    0.084    0.000   23.274   11.637 <frozen importlib._bootstrap>:659([_load_unlocked](http://127.0.0.1:4000/?func_name=_load_unlocked))
   3522/2    0.042    0.000   23.273   11.637 <frozen importlib._bootstrap_external>:844([exec_module](http://127.0.0.1:4000/?func_name=exec_module))
        1    0.000    0.000   23.271   23.271 /usr/local/airflow/astro_sdk/src/astro/databases/__init__.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000   23.224   23.224 /usr/local/airflow/astro_sdk/src/astro/databases/base.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
  1079/49    0.018    0.000   21.545    0.440 {built-in method builtins.__import__}
        1    0.000    0.000   17.856   17.856 /usr/local/lib/python3.9/site-packages/airflow/__init__.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000   16.472   16.472 /usr/local/lib/python3.9/site-packages/airflow/plugins_manager.py:283([ensure_plugins_loaded](http://127.0.0.1:4000/?func_name=ensure_plugins_loaded))
        1    0.000    0.000   16.465   16.465 /usr/local/lib/python3.9/site-packages/airflow/plugins_manager.py:217([load_entrypoint_plugins](http://127.0.0.1:4000/?func_name=load_entrypoint_plugins))
    82/18    0.001    0.000   15.412    0.856 /usr/local/lib/python3.9/importlib/__init__.py:109([import_module](http://127.0.0.1:4000/?func_name=import_module))
    91/18    0.001    0.000   15.412    0.856 <frozen importlib._bootstrap>:1018([_gcd_import](http://127.0.0.1:4000/?func_name=_gcd_import))
        4    0.000    0.000   13.975    3.494 /usr/local/lib/python3.9/site-packages/importlib_metadata/__init__.py:197([load](http://127.0.0.1:4000/?func_name=load))
        1    0.000    0.000   10.562   10.562 /usr/local/lib/python3.9/site-packages/astronomer/runtime/plugin.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
4051/1718    0.029    0.000    9.702    0.006 <frozen importlib._bootstrap>:1033([_handle_fromlist](http://127.0.0.1:4000/?func_name=_handle_fromlist))
        1    0.000    0.000    9.331    9.331 /usr/local/lib/python3.9/site-packages/astronomer/runtime/blueprint.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.001    0.001    9.300    9.300 /usr/local/lib/python3.9/site-packages/airflow/www/views.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
8912/8600    0.737    0.000    7.788    0.001 {built-in method builtins.__build_class__}
        1    0.000    0.000    3.254    3.254 /usr/local/lib/python3.9/site-packages/airflow/api/common/mark_tasks.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    3.234    3.234 /usr/local/lib/python3.9/site-packages/flask_appbuilder/__init__.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    3.191    3.191 /usr/local/lib/python3.9/site-packages/flask_appbuilder/api/__init__.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.001    0.001    3.175    3.175 /usr/local/lib/python3.9/site-packages/airflow/models/dag.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
     3522    0.155    0.000    3.013    0.001 <frozen importlib._bootstrap_external>:916([get_code](http://127.0.0.1:4000/?func_name=get_code))
        1    0.000    0.000    2.888    2.888 /usr/local/lib/python3.9/site-packages/pandas/__init__.py:3([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        5    0.017    0.003    2.374    0.475 /usr/local/lib/python3.9/site-packages/airflow/utils/entry_points.py:29([entry_points_with_dist](http://127.0.0.1:4000/?func_name=entry_points_with_dist))
       46    0.000    0.000    2.329    0.051 /usr/local/lib/python3.9/site-packages/airflow/utils/module_loading.py:23([import_string](http://127.0.0.1:4000/?func_name=import_string))
      371    0.012    0.000    2.242    0.006 /usr/local/lib/python3.9/site-packages/importlib_[metadata](http://127.0.0.1:4000/?func_name=metadata)/__init__.py:603(metadata)
        1    0.000    0.000    2.230    2.230 /usr/local/lib/python3.9/site-packages/flask_appbuilder/api/convert.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    2.211    2.211 /usr/local/lib/python3.9/site-packages/flask_appbuilder/models/sqla/interface.py:2([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    2.202    2.202 /usr/local/lib/python3.9/site-packages/airflow/models/operator.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.001    0.001    2.199    2.199 /usr/local/lib/python3.9/site-packages/airflow/models/baseoperator.py:18([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        5    0.000    0.000    2.197    0.439 /usr/local/lib/python3.9/site-packages/airflow/configuration.py:678([getimport](http://127.0.0.1:4000/?func_name=getimport))
        1    0.000    0.000    2.178    2.178 /usr/local/lib/python3.9/site-packages/airflow/models/baseoperator.py:438([BaseOperator](http://127.0.0.1:4000/?func_name=BaseOperator))
        1    0.000    0.000    2.175    2.175 /usr/local/lib/python3.9/site-packages/airflow/lineage/__init__.py:96([apply_lineage](http://127.0.0.1:4000/?func_name=apply_lineage))
        1    0.000    0.000    2.175    2.175 /usr/local/lib/python3.9/site-packages/airflow/lineage/__init__.py:43([get_backend](http://127.0.0.1:4000/?func_name=get_backend))
        1    0.000    0.000    2.172    2.172 /usr/local/lib/python3.9/site-packages/openlineage/lineage_backend/__init__.py:4([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
....
....
        1    0.000    0.000    0.000    0.000 /usr/local/lib/python3.9/_pyio.py:463([writable](http://127.0.0.1:4000/?func_name=writable))
        1    0.000    0.000    0.000    0.000 /usr/local/lib/python3.9/site-packages/google/cloud/bigquery_storage_v1/services/__init__.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    0.000    0.000 /usr/local/lib/python3.9/urllib/__init__.py:1([<module>](http://127.0.0.1:4000/?func_name=%3Cmodule%3E))
        1    0.000    0.000    0.000    0.000 /usr/local/lib/python3.9/site-packages/pyrsistent/_pmap.py:219([<lambda>](http://127.0.0.1:4000/?func_name=%3Clambda%3E))
```

**After**:

```
Thu Oct  6 23:16:05 2022    /Users/kaxilnaik/Documents/GitHub/astronomer/astro-sdk/python-sdk/dev/dags/program1.prof

         5751 function calls (5660 primitive calls) in 0.062 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     11/1    0.000    0.000    0.062    0.062 {built-in method builtins.exec}
     14/1    0.001    0.000    0.062    0.062 <frozen importlib._bootstrap>:1002(_find_and_load)
     14/1    0.000    0.000    0.061    0.061 <frozen importlib._bootstrap>:967(_find_and_load_unlocked)
     10/2    0.000    0.000    0.058    0.029 <frozen importlib._bootstrap>:659(_load_unlocked)
     10/2    0.000    0.000    0.057    0.029 <frozen importlib._bootstrap_external>:844(exec_module)
     14/3    0.000    0.000    0.055    0.018 <frozen importlib._bootstrap>:220(_call_with_frames_removed)
        1    0.000    0.000    0.049    0.049 /usr/local/airflow/astro_sdk/src/astro/databases/__init__.py:1(<module>)
```

---

**Cyclic Dep**:

**Before**:
```
❯ docker exec -it dev-airflow-scheduler-1 /bin/bash
astro@2d9566f3d1b0:/usr/local/airflow$ python
>>> from astro.databases import create_database
...
ImportError: cannot import name 'BaseDatabase' from 'astro.databases' (/usr/local/airflow/astro_sdk/src/astro/databases/__init__.py)
```

**After**:

```
❯ docker exec -it dev-airflow-scheduler-1 /bin/bash
astro@2d9566f3d1b0:/usr/local/airflow$ python
>>> from astro.databases import create_database
>>>
```

## Does this introduce a breaking change?
No

### Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: python-sdk/src/astro/__init__.py [hunk 1]
```diff
@@ -2,19 +2,6 @@
...
-#   from astro.sql.table import Metadata, Table, create_unique_table_name
-# We face ImportError, as it happened in:
-# https://github.com/astronomer/astro-sdk/pull/396/commits/fbe73bdbe46d65777258a5f79f461ef69f08a673
-# https://github.com/astronomer/astro-sdk/actions/runs/2378526135
-# Although astro.database does not depend on astro.sql, it depends on astro.sql.table - and, unless astro.sql was
-# imported beforehand, it will also load astro.sql. In astro.sql we import lots of operators which depend on
-# astro.database, and this is what leads to the circular dependency.
-
-import astro.sql  # noqa: F401
-
```
Comment (kaxil) [inline python-sdk/src/astro/__init__.py:17]: No longer needed

File: python-sdk/src/astro/databases/__init__.py [hunk 2]
```diff
@@ -1,10 +1,12 @@
 import importlib
 from pathlib import Path
+from typing import TYPE_CHECKING
 
-from airflow.hooks.base import BaseHook
-from astro.databases.base import BaseDatabase
```
Comment (kaxil) [inline python-sdk/src/astro/databases/__init__.py:5]: These two were the main cause of slowness !

`BaseDatabase` is just used for Type Hints !

File: python-sdk/src/astro/databases/__init__.py [hunk 3]
```diff
@@ -19,15 +21,17 @@
...
 
 
-def create_database(conn_id: str) -> BaseDatabase:
+def create_database(conn_id: str) -> "BaseDatabase":
     """
     Given a conn_id, return the associated Database class.
 
     :param conn_id: Database connection ID in Airflow
     """
+    from airflow.hooks.base import BaseHook
```
Comment (kaxil) [inline python-sdk/src/astro/databases/__init__.py:30]: This is only required for this function, hence imported lazily

File: python-sdk/src/astro/sql/table.py [hunk 4]
```diff
@@ -1,178 +1,8 @@
...
-        new_parsed_url = parsed_url._replace(query=urlencode(db_extra))
-        return new_parsed_url.geturl()
+from astro.table import (  # noqa: F401 # skipcq: PY-W2000
+    MAX_TABLE_NAME_LENGTH,
+    TEMP_PREFIX,
+    BaseTable,
+    Metadata,
+    Table,
+    TempTable,
+)
```
Comment (kaxil) [inline python-sdk/src/astro/sql/table.py:8]: This is for backwards-compatibility

File: python-sdk/src/astro/sql/__init__.py [hunk 5]
```diff
@@ -10,7 +10,7 @@
 from astro.sql.operators.merge import MergeOperator, merge
 from astro.sql.operators.raw_sql import RawSQLOperator, run_raw_sql
 from astro.sql.operators.transform import TransformOperator, transform, transform_file
-from astro.sql.table import Metadata, Table
+from astro.table import Metadata, Table
```
Comment (kaxil) [inline python-sdk/src/astro/sql/__init__.py:13]: This is for backwards-compatibility

**Review by feluelle — APPROVED** [review]
Nice 👍

**Pull request merged**